src/ui/public/navbar/navbar.js

```diff
@@ -0,0 +1,52 @@
+const _ = require('lodash');
+const $ = require('jquery');
+const navbar = require('ui/modules').get('kibana/navbar');
+
+require('ui/render_directive');
+
+navbar.directive('navbar', function (Private, $compile) {
+  const navbarExtensions = Private(require('ui/registry/navbar_extensions'));
+  const getExtensions = _.memoize(function (name) {
+    if (!name) throw new Error('navbar directive requires a name attribute');
+    return _.sortBy(navbarExtensions.byAppName[name], 'order');
+  });
+
+  return {
+    restrict: 'E',
+    template: function ($el, $attrs) {
+      const $buttonGroup = $el.children('.button-group');
+      if ($buttonGroup.length !== 1) throw new Error('navbar must have exactly 1 button group');
+
+      const extensions = getExtensions($attrs.name);
+      const controls = $buttonGroup.children().detach().toArray().map(function (button) {
```

Comment:
this would probably read cleaner if it was broken up a bit. Also, ***spread 'em!***
```js
const buttons = $buttonGroup.children().detach().toArray();
const controls = [
  ...buttons.map((button) => {
    return {
      order: 0,
      $el: $(button)
    };
  }),

  ...extensions.map((extension, index) => {
    return {
      extension,
      index,
      order: extension.order
    };
  })
]
```